In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Admin\Documents\Dec 22 Instacart Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [7]:
#practice dataframe
df_test = pd.DataFrame()

In [8]:
#practice dataframe columns
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
#looking for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [5]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [6]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [ ]:
df_pr

In [19]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task

## step 2

In [20]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

I want to know what is so expensive that it costs $99999 lol. It's probably only one item because the mean is still very low, just likely not representative of what's actually going on in that column. 

You can see the 5 duplicates in this original product set because the count value is higher than the max value and in this context that indicates a problem

## step 3

In [22]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

I guess there's no columns with mixed data types? Lol feels like there should be one for the purpose of this task. If there were one, we'd decide the appropriate data type and use the astype function to fix it. 

## step 5

In [23]:
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In this case we shouldn't delete the missing rows. Days since last order will only have a value after they've placed their first order, so an empty value here just means it's their first order. 

## step 7

In [24]:
#calling it duplicates since we already used dups in the exercise
df_duplicates = df_ords[df_ords.duplicated()]

In [25]:
df_duplicates

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

no duplicates to be found here!